[mlir][cmake] Do not export MLIR_MAIN_SRC_DIR and MLIR_INCLUDE_DIR

nikic opened the pull request:
MLIR_MAIN_SRC_DIR and MLIR_INCLUDE_DIR point to the source directory, which is not installed. As such, the installed MLIRConfig.cmake also should not reference it.

The comment indicates that these are needed for mlir_tablegen(), but I don't see any related uses.

The motivation for this is the use in flang, where we end up inheriting a meaningless MLIR_MAIN_SRC_DIR from a previous MLIR build, whose source directory doesn't exist anymore, and that cannot be overridden with the correct path, because it's not a cached variable.

Instead do what all the other projects do for LLVM_MAIN_SRC_DIR and initialize MLIR_MAIN_SRC_DIR to CMAKE_CURRENT_SOURCE_DIR/../mlir.

For MLIR_INCLUDE_DIR there already is an exported MLIR_INCLUDE_DIRS, which can be used instead.

nikic committed:
[mlir][cmake] Do not export MLIR_MAIN_SRC_DIR

MLIR_MAIN_SRC_DIR points to the source directory, which is not
installed. As such, the installed MLIRConfig.cmake also should
not reference it.

The comment indicates that this is needed for mlir_tablegen(),
but I don't see any related uses.

The motivation for this is the use in flang, where we end up
inheriting a meaningless MLIR_MAIN_SRC_DIR from a previous MLIR
build, whose source directory obviously doesn't exist anymore,
and that cannot be overridden with the correct path, because it's
not a cached variable.

Instead do what all the other projects do (for LLVM_MAIN_SRC_DIR)
and initialize MLIR_MAIN_SRC_DIR to CMAKE_CURRENT_SOURCE_DIR/../mlir.

joker-eph reviewed (approved):
I don't quite understand why the line you removed in `mlir/cmake/modules/MLIRConfig.cmake.in` was there, but if it break something we can always revert an document better the setup.

What about the line before you left: `set(MLIR_INCLUDE_DIR "@MLIR_INCLUDE_DIR@")` ?

nikic commented:
> What about the line before you left: `set(MLIR_INCLUDE_DIR "@MLIR_INCLUDE_DIR@")` ?

Good point. It also points to the include directory in the source tree, so it should be dropped as well.

nikic committed:
Also drop MLIR_INCLUDE_DIR

nikic commented:
I've updated this to drop MLIR_INCLUDE_DIR now. Note that there already is MLIR_INCLUDE_DIRS, which is set up correctly depending on whether it's the installed or the build-tree config.

I've dropped the MLIR_INCLUDE_DIR use for FIROps.td because I don't think it's actually necessary. We already add the MLIR_INCLUDE_DIRS to the include directories at a higher level, so we don't need to do so again here. (Otherwise I could switch this to use EXTRA_INCLUDES with MLIR_INCLUDE_DIRS.)

nikic commented:
@joker-eph Would you mind taking another look at this (now that it also removes MLIR_INCLUDE_DIR)?

joker-eph reviewed (approved):
Sorry, I had approve before and thought it was enough, but happy to re-approve explicitly of course. Thanks.

nikic commented:
/cherry-pick 82bd148a3f25439d7f52a32422dc1bcd2da03803
